rust/hyperlane-base/src/db/rocks/hyperlane_db.rs

```diff
@@ -152,6 +154,23 @@ impl HyperlaneRocksDB {
         Ok(true)
     }
 
+    /// Store the merkle tree insertion event, and also store a mapping from message_id to leaf_index
+    pub fn process_tree_insertion(&self, insertion: &MerkleTreeInsertion) -> DbResult<bool> {
+        if let Ok(Some(_)) = self.retrieve_merkle_leaf_index_by_message_id(&insertion.message_id())
+        {
+            debug!(insertion=?insertion, "Tree insertion already stored in db");
+            return Ok(false);
+        }
+        // even if double insertions are ok, store the leaf by `leaf_index` (guaranteed to be unique)
+        // rather than by `message_id` (not guaranteed to be unique), so that leaves can be retrieved
+        // based on insertion order.
+        self.store_merkle_tree_insertion_by_leaf_index(&insertion.index(), insertion)?;
```

Comment:
where are we checking for uniqueness per leaf_index?

shouldn't `retrieve_merkle_message_id_by_leaf_index(leaf_index) == 0 || retrieve_merkle_message_id_by_leaf_index(leaf_index) == insertion.message_id()` be necessary too here?